File: app/javascript/map/geocoder.js
// ================================================================
// Geocoder（単一責務）
// 用途:
// - 住所文字列を Geocoding して位置情報を取得する
// - 失敗時は Promise を reject して呼び出し元にエラーを返す
// ================================================================

export const geocodeAddress = (address) => {
  // 呼び出し元で await できるよう Promise 化
  return new Promise((resolve, reject) => {

    // 1) 入力チェック（空なら失敗として返す）
    if (!address || address.trim() === "") {
      // ✅ reject = 「失敗として返す」(表示するかは呼び出し元の責務)
      reject(new Error("住所が空です"));
      return;
    }

    // 2) Google Maps API が読み込まれているかチェック
    if (!window.google?.maps?.Geocoder) {
      reject(new Error("Google Maps API (Geocoder) が読み込まれていません"));
      return;
    }

    // 3) Geocoder を作成して、住所 → 座標変換を実行
    const geocoder = new google.maps.Geocoder();

    geocoder.geocode({ address: address.trim() }, (results, status) => {

      // 4) 失敗判定（status がOKじゃない / 結果が空）
      if (status !== "OK" || !results || results.length === 0) {
        reject(new Error(`Geocoding に失敗しました: ${status}`));
        return;
      }

      // 5) 成功：1件目を採用して必要な情報だけ返す
      const result = results[0];

      resolve({
        location: result.geometry.location,     // google.maps.LatLng（lat(), lng() が取れる）
        viewport: result.geometry.viewport,     // google.maps.LatLngBounds | undefined（fitBoundsに使える）
        formattedAddress: result.formatted_address, // 表示用住所（例: "日本、〒... 栃木県..."）
        placeId: result.place_id,               // Google Places ID
      });
    });
  });
};

// ================================================================
// 表示用の住所整形（単一責務）
// 用途:
// - Googleが返す formatted_address は長いので、表示に不要な部分を削る
//   例: "日本、〒xxx-xxxx 栃木県..." → "栃木県..."
// ================================================================

export const normalizeDisplayAddress = (formattedAddress) => {
  // ✅ 何もなければ空文字
  if (!formattedAddress) return "";

  // 1) 先頭の「日本、」を削除
  let s = formattedAddress.replace(/^日本、\s*/u, "");

  // 2) 先頭の郵便番号「〒123-4567 」を削除
  s = s.replace(/^〒\d{3}-\d{4}\s*/u, "");

  // 3) 前後の空白を削って返す
  return s.trim();
};

File: app/clients/google_api/geocoder.rb
# frozen_string_literal: true

# 責務: Google Geocoding API を呼び出し、住所と座標を相互変換する
#
# 提供メソッド:
#   - forward: 住所から緯度経度を取得
#   - reverse: 緯度経度から住所を取得
#
# 注意: エラー時はnilを返す。フォールバック値はドメイン層で決定すること。
#
class GoogleApi::Geocoder
  API_URL = "https://maps.googleapis.com/maps/api/geocode/json"

  class << self
    # 住所から緯度経度を取得（Forward Geocoding）
    # @param address [String] 住所文字列
    # @return [Hash, nil] { lat:, lng:, prefecture: } or nil
    def forward(address)
      return nil if address.blank?

      uri = build_uri(address: address)
      json = GoogleApi.fetch_json(uri)

      if json["status"] == "OK"
        result = json["results"].first
        location = result.dig("geometry", "location")
        components = result["address_components"] || []

        prefecture = components.find { |c| c["types"].include?("administrative_area_level_1") }&.dig("long_name")

        { lat: location["lat"], lng: location["lng"], prefecture: prefecture }
      else
        Rails.logger.warn "[GoogleApi::Geocoder] forward failed: #{json['status']}"
        nil
      end
    rescue StandardError => e
      Rails.logger.error "[GoogleApi::Geocoder] forward error: #{e.message}"
      nil
    end

    # 緯度経度から住所を取得（Reverse Geocoding）
    # @param lat [Float] 緯度
    # @param lng [Float] 経度
    # @return [Hash, nil] { lat:, lng:, address:, prefecture:, city:, town: } or nil
    def reverse(lat:, lng:)
      return nil unless lat.present? && lng.present?

      uri = build_uri(latlng: "#{lat},#{lng}")
      json = GoogleApi.fetch_json(uri)

      if json["status"] == "OK"
        result = json["results"].first
        components = result["address_components"] || []

        {
          lat: lat,
          lng: lng,
          address: GoogleApi.normalize_address(result["formatted_address"].to_s),
          prefecture: extract_component(components, "administrative_area_level_1"),
          city: extract_component(components, "locality", "sublocality_level_1"),
          town: extract_component(components, "sublocality_level_2", "sublocality_level_3", "neighborhood")
        }
      else
        Rails.logger.error "[GoogleApi::Geocoder] reverse failed: #{json['status']}"
        nil
      end
    rescue StandardError => e
      Rails.logger.error "[GoogleApi::Geocoder] reverse error: #{e.message}"
      nil
    end

    private

    def build_uri(params)
      uri = URI.parse(API_URL)
      uri.query = URI.encode_www_form(
        params.merge(
          key: GoogleApi.api_key,
          language: "ja",
          region: "jp"
        )
      )
      uri
    end

    def extract_component(components, *types)
      components.find { |c| (c["types"] & types).any? }&.dig("long_name")
    end
  end
end
